### ISSUE TITLE
Helm Hook "Post Install" not working for Dependencies Chart

### ISSUE THREAD

**Issue by robin-carry** [helm/helm#11422]
https://helm.sh/docs/topics/charts_hooks/#writing-a-hook - Trying to get this working

**Requirements**: I want the `console` pod to start at the end in the below Chart X which has dependency to `console` chart.

**Solution Space**: Thought of the helm hooks as a life-cycle to deploy it at the end.

**Result/Behavior**: The `annotations` getting applied, but the `console` pods starts as soon as I deploy the whole chart.

**Problem**: Because Console is the UI for kafka cluster, it needs to wait for kafka broker to be available. I want to run it post-install of a Helm Chart installed. But it doesn’t…

What am I doing wrong? Or is it not a supported feature?

<img width="597" alt="Screen Shot 2022-10-06 at 9 48 18 AM" src="https://user-images.githubusercontent.com/4965919/194622476-e62eea83-788d-4185-8db4-878b4d7a3c75.png">
<img width="640" alt="Screen Shot 2022-10-06 at 9 48 58 AM" src="https://user-images.githubusercontent.com/4965919/194622483-a7c9acd2-68b7-4f6a-b2af-e086960f4cb7.png">
<img width="658" alt="Screen Shot 2022-10-06 at 9 49 41 AM" src="https://user-images.githubusercontent.com/4965919/194622484-cb5a9209-0b96-40be-9c7f-ce3e10d7bdf6.png">

Output of `helm version`: 
```bash
version.BuildInfo{Version:"v3.10.0", GitCommit:"ce66412a723e4d89555dc67217607c6579ffcb21", GitTreeState:"clean", GoVersion:"go1.19.1"}
```

Output of `kubectl version`:
```bash
kubectl version                    
WARNING: This version information is deprecated and will be replaced with the output from kubectl version --short.  Use --output=yaml|json to get the full version.
Client Version: version.Info{Major:"1", Minor:"25", GitVersion:"v1.25.2", GitCommit:"5835544ca568b757a8ecae5c153f317e5736700e", GitTreeState:"clean", BuildDate:"2022-09-21T14:25:45Z", GoVersion:"go1.19.1", Compiler:"gc", Platform:"darwin/amd64"}
Kustomize Version: v4.5.7
Server Version: version.Info{Major:"1", Minor:"25", GitVersion:"v1.25.0", GitCommit:"a866cbe2e5bbaa01cfd5e969aa3e033f3282a8a2", GitTreeState:"clean", BuildDate:"2022-08-23T17:38:15Z", GoVersion:"go1.19", Compiler:"gc", Platform:"linux/amd64"}
```

Cloud Provider/Platform (AKS, GKE, Minikube etc.): 
1. Docker-Desktop
2. Rancher on Azure VM

PS: Unable to add label (not sure why) - this in my view is `question/support` type of ticket

**Comment by joejulian** [comment]
I'd like to take this question a little sideways, if I may.

> Because Console is the UI for Kafka cluster, it needs to wait for Kafka broker to be available
Why?

What happens if the broker isn't yet available? The console fails and the pod exits. The Kubernetes scheduler starts the pod again. It may fail again. If it does, it's retried (with a graceful fallback) until it hits the maximum of 5 minutes.

As long as you're not reaching the CrashLoopBackoff, I'd just consider letting it fail a few times. If you are, I'd probably add some retries in the start command or a sleep if the error is being unable to connect to the brokers.

In other words I'm suggesting that you let Kubernetes handle the scheduling and let your helm chart be declarative.

**Comment by robin-carry** [comment]
@joejulian - What you have said has the absolute truth. That is what I am doing right now, but that is not what I was asking. I was asking about the behavior of `helm` - where the `post-install` hook wasn't working. 

This may be the example I am facing right now, but in future I might have distributed services with mini-liths where restarts/failure is waste (due to legacy challenges). 

While this current problem works, I still didn't learn for what reason `helm` didn't work as it is documented (or how I have understood that document). And I will raise the same question later when the example comes around. 

Question -> Did I do something wrong? Or `post-install` is not supposed to be used this way? Or in this specific case the `post-install` just didn't work/not supported feature as of now.

**Comment by joejulian** [comment]
Are you using the `--wait` flag? It looks like the hooks will only wait until the main body is ready if that flag is used: https://github.com/helm/helm/blob/f08db5c899c4e312c72ae863acbb90ac8a4747ed/pkg/action/install.go#L356-L420

**Comment by robin-carry** [comment]
@joejulian - I didn't earlier (when I reported the issue), but I just did and still same result. 

If you want to experiment, I have it all in my private git - https://github.com/robin-carry/strimzi-parent 

```bash
git clone https://github.com/robin-carry/strimzi-parent
cd strimzi-parent
./0-init.sh
helm install my-cluster . --wait
```

I still see immediately `my-cluster-console-****` pod started and running, trying to connect to the broker endpoint, which isn't available.

**Comment by joejulian** [comment]
I mean... I like the console you're using, but the rest of it's so big and slow... :wink:

**Comment by joejulian** [comment]
Ok, you only have 2 resources that can be waited for, both Deployments. The `strimzi-cluster-operator` installs and is ready very quickly, which allows helm to move on to the hooks. There's no way for helm to know how to wait for a `Kafka` or `KafkaConnect` resource so it doesn't.

**Comment by joejulian** [comment]
If you installed your kafka using a kubernetes primitives that have known readiness checks, then you could use hooks. Like if you used the [Redpanda helm chart](https://github.com/redpanda-data/helm-charts), which uses a StatefulSet, helm knows how to wait for that to be ready so your hook could then run after it is.

**Comment by robin-carry** [comment]
> Ok, you only have 2 resources that can be waited for, both Deployments. The `strimzi-cluster-operator` installs and is ready very quickly, which allows helm to move on to the hooks. There's no way for helm to know how to wait for a `Kafka` or `KafkaConnect` resource so it doesn't.

Did I understand you correctly - Either the resource has to be of `kind: deployment` or because my resources which are getting deployed have no `probes` defined, that is why the `helm-hooks` can't wait for them as they can't identify?

If either of the above understanding is true - is it possible to update the documentation too with it (I might do it too, but not a native english speaker & clearly not an expert here).

**Comment by joejulian** [comment]
These are the resources that helm knows how to wait for:
https://github.com/helm/helm/blob/main/pkg/kube/wait.go#L109-L145

If they're not in that list, no waiting is done. You can't really wait for a configmap to be ready or a secret, and the same is true for any custom resource. There's just no standard that's enforced by the kubernetes api server.

**Comment by robin-carry** [comment]
Above ⬆️ reply clarifies where I was going wrong. So custom-crds are not in scope of helm-hooks I guess. 

Can this be a feature request? Wait for `crd` of type `x` for `helm-hooks`?

**Comment by joejulian** [comment]
That would need to go into k8s first. They'd need to provide a standard for communicating how a custom resource's status would be considered "ready".

**Comment by robin-carry** [comment]
Cool... I would say this question has been answered. And I have understood the challenge behind it. As of now there isn't any way - so I will depend on the `k8s-scheduler` for a failing pod to killed/restarted.

**Comment by prashathsenthil** [comment]
@robin-carry I'm facing a similar issue, so the resolution is that when helm-hook annotations are added to the resources that uses CRD then at this time its not supported... is that a correct understanding?

I have a resource that uses a CRD to create ClusterRole and ClusterRoleBinding, basically I want this to get deployed first, so I added pre-install, pre-upgrade hooks hoping that it go's in first followed by my other dependency chart resources, however its not working i.e. the  ClusterRole and ClusterRoleBinding aren't getting deployed via the hook, saw this post which was along the same lines, so thought of getting some clarification.

**Comment by robin-carry** [comment]
>@robin-carry I'm facing a similar issue, so the resolution is that when helm-hook annotations are added to the resources that uses CRD then at this time its not supported... is that a correct understanding?

It is not like it is `not supported`. It is like `it can't be supported` as of now, because `k8s` gives no way to understand when a custom resource can be marked ready. And hence the current implementation of `helm` can only identify limited set of `k8s-resources` for which `Helm-Hooks` are applicable. 

From the [code base](https://github.com/helm/helm/blob/main/pkg/kube/wait.go#L109-L145), we can see that only handful few resources can be - namely `Deployment`, `ReplicationSet`, `ReplicationController, `StatefulSet`, `DaemonSet`, `Service` & `Job`.

**Comment by prashathsenthil** [comment]
> > @robin-carry I'm facing a similar issue, so the resolution is that when helm-hook annotations are added to the resources that uses CRD then at this time its not supported... is that a correct understanding?
> 
> It is not like it is `not supported`. It is like `it can't be supported` as of now, because `k8s` gives no way to understand when a custom resource can be marked ready. And hence the current implementation of `helm` can only identify limited set of `k8s-resources` for which `Helm-Hooks` are applicable.
> 
> From the [code base](https://github.com/helm/helm/blob/main/pkg/kube/wait.go#L109-L145), we can see that only handful few resources can be - namely `Deployment`, `ReplicationSet`, `ReplicationController, `StatefulSet`, `DaemonSet`, `Service`&`Job`.
> 
> Your request is about `pre-install`, while I am no expert of the `helm-hooks` I presume it could be the same reason too.

Thank you, it makes sense. Appreciate the quick response
